File: SAI_RIDEPROCESSORJAVA/RideProcessor.java
import java.util.*;
import java.util.concurrent.locks.*;
import java.util.concurrent.*;
import java.util.logging.*;

public class RideProcessor {
    private static final Logger logger = Logger.getLogger(RideProcessor.class.getName());
    private static final Queue<String> taskQueue = new LinkedList<>();
    private static final List<String> results = new ArrayList<>();
    private static final Lock queueLock = new ReentrantLock();
    private static final Lock resultLock = new ReentrantLock();

    public static void main(String[] args) {
        int numWorkers = 3;
        Thread[] workers = new Thread[numWorkers];

        // Populate task queue
        for (int i = 1; i <= 12; i++) {
            taskQueue.add("RideRequest-" + i);
        }

        // Start workers
        for (int i = 0; i < numWorkers; i++) {
            workers[i] = new Thread(new Worker(i));
            workers[i].start();
        }

        // Wait for completion
        for (Thread worker : workers) {
            try {
                worker.join();
            } catch (InterruptedException e) {
                logger.log(Level.SEVERE, "Thread interrupted", e);
            }
        }

        // Output results
        System.out.println("\nFinal Results:");
        for (String result : results) {
            System.out.println(result);
        }
    }

    static class Worker implements Runnable {
        private final int id;

        Worker(int id) {
            this.id = id;
        }

        public void run() {
            logger.info("Worker " + id + " started.");
            while (true) {
                String task = null;

                queueLock.lock();
                try {
                    if (!taskQueue.isEmpty()) {
                        task = taskQueue.poll();
                    }
                } finally {
                    queueLock.unlock();
                }

                if (task == null) break;

                try {
                    Thread.sleep(400); // Simulate processing
                    String result = "Worker " + id + " completed " + task;

                    resultLock.lock();
                    try {
                        results.add(result);
                    } finally {
                        resultLock.unlock();
                    }

                    logger.info(result);
                } catch (InterruptedException e) {
                    logger.log(Level.WARNING, "Worker " + id + " interrupted", e);
                } catch (Exception e) {
                    logger.log(Level.SEVERE, "Worker " + id + " error", e);
                }
            }
            logger.info("Worker " + id + " finished.");
        }
    }
}

File: SAI_RIDEPROCESSORGO/ride_processor.go
package main

import (
	"fmt"
	"log"
	"os"
	"sync"
	"time"
)

type TaskQueue struct {
	tasks []string
	lock  sync.Mutex
}

func (q *TaskQueue) GetTask() (string, bool) {
	q.lock.Lock()
	defer q.lock.Unlock()
	if len(q.tasks) == 0 {
		return "", false
	}
	task := q.tasks[0]
	q.tasks = q.tasks[1:]
	return task, true
}

func main() {
	logFile, err := os.OpenFile("ride_processor.log", os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0666)
	if err != nil {
		log.Fatalf("Log file error: %v", err)
	}
	defer logFile.Close()
	log.SetOutput(logFile)

	queue := &TaskQueue{tasks: make([]string, 0)}
	for i := 1; i <= 12; i++ {
		queue.tasks = append(queue.tasks, fmt.Sprintf("RideRequest-%d", i))
	}

	var results []string
	var resultLock sync.Mutex
	var wg sync.WaitGroup
	numWorkers := 3

	for i := 0; i < numWorkers; i++ {
		wg.Add(1)
		go processWorker(i, queue, &results, &resultLock, &wg)
	}

	wg.Wait()

	fmt.Println("\nFinal Results:")
	for _, r := range results {
		fmt.Println(r)
	}
}

func processWorker(id int, queue *TaskQueue, results *[]string, lock *sync.Mutex, wg *sync.WaitGroup) {
	defer wg.Done()
	log.Printf("Worker %d started.", id)

	for {
		task, ok := queue.GetTask()
		if !ok {
			break
		}

		time.Sleep(400 * time.Millisecond)
		result := fmt.Sprintf("Worker %d completed %s", id, task)

		lock.Lock()
		*results = append(*results, result)
		lock.Unlock()

		log.Println(result)
	}

	log.Printf("Worker %d finished.", id)
}
